creusot/tests/should_succeed/ghost/pcell.rs

```diff
@@ -0,0 +1,158 @@
+extern crate creusot_contracts;
+use creusot_contracts::*;
+
+pub struct Id(()); // TODO: use a why3 abstract type instead ?
+
+impl Id {
+    /// Generate a fresh id.
+    ///
+    /// To ensure that it is different from previously generated ids, see [`Self::differentiate`].
+    #[pure]
+    #[requires(true)]
+    #[ensures(true)]
+    fn fresh() -> Self {
+        Self(())
+    }
+
+    /// When creating two `PCell`s, we don't know that their ids are different.
+    ///
+    /// This function allows us to derive exactly that.
+    #[trusted]
+    #[pure]
+    #[ensures(*self == ^self)]
+    #[ensures(^self != *other)]
+    fn differentiate(&mut self, other: &Self) {}
+}
```

Comment:
There is no way this is sound. You definitely need the Id to be an infinitely sized type. I should be able to prove that `Id::fresh() == Id::fresh()`. You can put `#[trusted]` on the `Id` but I think it should be a wrapper around an `Int`.

Reply:
Indeed, although `Id` does not need to be infinitely sized, only to contain more than one value (e.g. `bool` is enough). I think @jhjourdan was unsure about `Int`, as it could give the solvers to (wrong) idea to use arithmetic ?

Reply:
`()` and `bool` are not better than `Int`, they are all concrete types. I proposed to use an *abstract* type here (i.e., using `#[trusted]` on the type), so that solvers only have the theory of equality on this type.

Reply:
Yea, I think that would be best, though we should ideally come up with a better way of doing that.

Reply:
Once we will honor `pub` annotations on fields of structures, that will be easy.

Reply:
All right, am I right to put `#[trusted]` on `Id::fresh` then ?

Reply:
no on the `Id` type itself. This causes creusot to translate it to an uninterpreted type.